### ISSUE TITLE
Deep dive discussion on {Async,Promise}_hooks & performance

### ISSUE THREAD

**Issue by ofrobots** [nodejs/diagnostics#248]
During the WG meeting today we discussed that It would be good to meet and unblock the topic of async hooks / promise hooks performance. (https://github.com/nodejs/diagnostics/issues/188).

Context:
* [Proposal by @hashseed](https://docs.google.com/document/d/1y4lF-iQhhuSOgzlPIBbfCT4IUpg-in3az5Q87C7aY3Q/edit)
* https://github.com/nodejs/node/pull/21313

Here's a doodle: https://doodle.com/poll/fqkzvw929ueys2u9 I want to be able to get at least @hashseed @mcollina @AndreasMadsen to be able to attend. Please mark your availability by Friday Oct 19 and I will get a deep dive discussion scheduled.

**Comment by AndreasMadsen** [comment]
I would like @mike-kaufman or another from Microsoft to also attend, as they appear to be the primary driver of "JavaScript Async Context".

**Comment by Qard** [comment]
I'd like to attend if I can, so I put my times up. I have some thoughts on the subject. It's okay if it doesn't line up for me though. Definitely @hashseed, @mcollina and @AndreasMadsen are the core people we need most in the conversation.

**Comment by ofrobots** [comment]
It would be great to have as many people attend as possible. Please add your preferences, and I will try my best to find a good timeslot 🤞.

**Comment by Flarna** [comment]
These evening slots are in general hard for me and also difficult to predict. I don't add anything in doodle to avoid to bias it but I will try to attend.

**Comment by davisjam** [comment]
Will this conversation be recorded?

**Comment by mhdawson** [comment]
I'll try to attend as well, but think it is more important that times are good for other participants so won't complete the doodle.

**Comment by ofrobots** [comment]
Based on the poll, the best meeting time slot seems to be Oct 30, 11am PDT (1900 Central Europe) ([your local time](https://www.wolframalpha.com/input/?i=11am+PDT+Oct+30+in+my+timezone)).

Meeting link: https://meet.google.com/tov-fcdh-jfw

**Comment by AndreasMadsen** [comment]
Is that 18:00 UTC? I'm switching to daylight saving time on the 28th, so I'm a bit worried and would like a fixed timezone specifier.

**Comment by wentout** [comment]
I'm very sorry, but I'm very interested in the result of this discussion.
Can I listen also?
I closed [Issue #249]https://github.com/nodejs/diagnostics/issues/249), cause seems is was bit duplicate and everyone knows the idea, described there.

**Comment by ofrobots** [comment]
@AndreasMadsen AFAIK that [should be](https://www.wolframalpha.com/input/?i=11am+PDT+Oct+30+in+UTC) Oct 30th 1800 UTC. Timezones are hard 😂.

@wentout you're welcome to join the hangouts at the time of the meeting!

**Comment by wentout** [comment]
![image](https://user-images.githubusercontent.com/1651439/47479854-0b594200-d837-11e8-9e42-3d57b73482d6.png)

**Comment by mmarchini** [comment]
I'm on the road and won't make it :(

**Comment by mcollina** [comment]
Here is the link to the domain issue: https://github.com/nodejs/node/issues/23862

**Comment by mcollina** [comment]
Here is the link to the `currentAsyncResource()` latest PR: https://github.com/mcollina/node/tree/current-resource-bis.

**Comment by ofrobots** [comment]
The recording of the meeting is here: https://youtu.be/2SCNsozMkF8. Thanks for everyone for participating.

Here's a brief summary of the action plan:
* We discussed that it would be great if majority of the async-hooks uses-cases didn't need to use the destroy hook. The hook can stay, but the expectation is that the availability of the resource on hook callbacks will remove the need for APMs to install destroy hooks. @mcollina's PR nodejs/node#21313 needs to be resurrected. Looking for volunteers. If there are no volunteers, @mcollina will pick this up again once he has some free time in December. There needs to be some consideration of whether it would problematic to expose the promise as the resource object. @hashseed to check with the V8 compiler team about gotchas. @ofrobots and @AndreasMadsen to provide concrete examples where we observe promises taking too long to GC.
* @hashseed to work with Maya on the V8 compiler team to introduce promise hooks from JS. Once these are added, Node core can switch to these.
* General agreement that non-observable promises don't need to trigger hooks as long as context preserved. V8 team to implement this optimization.
* V8 team expressed a desire for some spec modification describing promise hooks behavior.  General agreement that this would be a good thing to get the interface between V8 and Node solidified. @hashseed to point to specific section and then WG can work drafting some spec test.

Let me know if I missed or misunderstood anything. Full details in the recording.

**Comment by mmarchini** [comment]
> @mcollina's PR nodejs/node#21313 needs to be resurrected. Looking for volunteers. If there are no volunteers

I can do that. Should I just resubmit the same PR or are there any changes required before resubmitting it?

**Comment by mcollina** [comment]
@mmarchini there are some comments in the PR that needs to be fixed, mainly there is an ask that we maintain a stack of the current async resources.

**Comment by mmarchini** [comment]
Ok, thanks. I'll take a look into it next week.

**Comment by Flarna** [comment]
@mcollina Could you please point to the fix regarding `http_parser` which allows to use the async resource in before/after?

**Comment by mcollina** [comment]
@Flarna here it is https://github.com/nodejs/node/pull/23201.

**Comment by AndreasMadsen** [comment]
@mcollina That PR looks like it is just a small fix. As far as I can see, the same resource object is still being reused. I would expect a fix that allows `WeakMap` to be used, to involve something with the `freelist` and the HTTP_PARSER allocator.

**Comment by mcollina** [comment]
@AndreasMadsen I cannot reproduce the problem anymore. Can you share a snippet that reproduces it?

```js
'use strict'

const sleep = require('util').promisify(setTimeout)
const { createHook, executionAsyncResource } = require('async_hooks')
const asyncHook = createHook({init})
const sym = Symbol('cls')

asyncHook.enable()

function getCLS () {
  return executionAsyncResource()[sym]
}

function setCLS (data) {
  return executionAsyncResource()[sym] = data
}

function init (asyncId, type, triggerAsyncId, resource) {
  var cr = executionAsyncResource()
  if (cr) {
    if (type === 'HTTP_PARSER') {
      if (cr.noReuse) {
        throw new Error('no reuse')
      }

      cr.noReuse = true
    }

    resource[sym] = cr[sym]
  }
}


var counter = 0

const server = require('http').createServer(function (req, res) {
  setCLS(counter++)
  sleep(10).then(function () {
    res.setHeader('content-type', 'application/json')
    res.end(JSON.stringify({ cls: getCLS() }))
  })
})
server.listen(3000)
```

**Comment by AndreasMadsen** [comment]
@mcollina I'm not in a position to compile anything right now, but

```js
'use strict'

const http = require('http')
const { createHook } = require('async_hooks')
const reused = Symbol('reused')

const asyncHook = createHook({
  init (asyncId, type, triggerAsyncId, resource) {
    if (resource[reused]) {
      throw new Error('reuse detected')
    }
    resource[reused] = true;
  }
})
asyncHook.enable()

const server = http.createServer(function (req, res) {
  res.end();
})
server.listen(3000, function () {
  http.get('http://127.0.0.1:3000', function () {
    http.get('http://127.0.0.1:3000', function () {
      console.log('passed');
      server.close();
    });
  });
});
```

is the test you need.

**Comment by AndreasMadsen** [comment]
@mcollina tested it now, the issue is still there.

**Comment by mcollina** [comment]
What I checked is the example on https://github.com/nodejs/diagnostics/issues/248#issuecomment-436236180. Before the PR I linked, that example was leaking memory. After, it does not. The reason for  that is that in normal http usage, there is going to be no context attached to `http_parser`. So, not leak.

I agree we should fix this. @mmarchini would you still have time to tackle this?

**Comment by mmarchini** [comment]
Yes, I'll have some time tomorrow and Friday to work on this.

**Comment by mmarchini** [comment]
Ok, I rebased nodejs/node#21313 and made some changes to store the resource in a stack. The code still needs some cleanup before opening a new PR though. I also want to run benchmarks to make sure there's no regression using this approach. If there's no significant regressions I'll open the PR tomorrow. If anyone is interested, the code is here: https://github.com/mmarchini/node/tree/current-resource.

> Maybe some dummy/wrapper object (similar as for Promises) should be created on every alloc and passed as resource to the callbacks.

I believe this could be implemented in a separate Pull Request, right?

**Comment by AndreasMadsen** [comment]
@mmarchini I'm going to ❌ it, until https://github.com/nodejs/diagnostics/issues/248#issuecomment-436252982 has been fixed. Otherwise it is just logically invalid. I hope you understand.

**Comment by Flarna** [comment]
> I believe this could be implemented in a separate Pull Request, right?

I think so but the sequence should be the to have unique HTTPParser resources first. But on the other hand that proxy/wrapper is only needed if we really go for this solution otherwise it's just needless overhead.
To my understanding at least the memory performance of using a weakmap needs to be checked before we can apply this change.

**Comment by mmarchini** [comment]
Implementing a wrapper on HTTPParser will be tricky 🤔. Do we really need to reuse HTTPParsers though? I changed the max amount of parsers in the FreeList at to 0, which forces Node.js to always create a new parser, and our benchmarks don't show significant regressions:

```
                                      confidence improvement accuracy (*)   (**)  (***)
 http/bench-parser.js n=100000 len=16                -2.47 %       ±3.05% ±4.07% ±5.34%
 http/bench-parser.js n=100000 len=32                -1.58 %       ±3.74% ±4.98% ±6.48%
 http/bench-parser.js n=100000 len=4                 -0.02 %       ±4.74% ±6.31% ±8.22%
 http/bench-parser.js n=100000 len=8                 -0.94 %       ±3.44% ±4.58% ±5.96%

Be aware that when doing many comparisons the risk of a false-positive
result increases. In this case there are 4 comparisons, you can thus
expect the following amount of false-positive results:
  0.20 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.04 false positives, when considering a   1% risk acceptance (**, ***),
  0.00 false positives, when considering a 0.1% risk acceptance (***)
```

![compare-plot](https://user-images.githubusercontent.com/4048656/48292582-c08f1b00-e42f-11e8-858d-a38f762a5be5.png)


(I might be running the wrong benchmarks though. Are there any other recommended benchmarks I should run?)

**Comment by mcollina** [comment]
Why don’t you send a PR that removes that reuse? Then we can run our
benchmark ci.

I think it might not be needed anymore because: a) we have a new GC b)
referencing from “old” objects to “new” objects is costly, and this reuse
was causing that. In an ideal world, reusing that should not be needed.

Il giorno sab 10 nov 2018 alle 00:02 Matheus Marchini <
notifications@github.com> ha scritto:

**Comment by mmarchini** [comment]
That's what I was thinking. Also, I don't think there would be any difference between removing the reuse and wrapping HTTPParse (we would need to allocate the wrapper every time anyway).

**Comment by Flarna** [comment]
I think the main difference is that a wrapper object would be a simple javascript object linked to HTTPParser and this link is cut once the parser is moved into the freelist whereas a HTTPParser object has a native part attached so it's more heavy.
But I fully agree that it's even better to avoid the complexity of introducing a wrapper just because of this.

**Comment by Flarna** [comment]
Besides the reuse of `HTTPParser` I found two more places for reuse in node core:
* `HttpAgent` reuses it's handle
* `FileHandle::ReadStart()` reuses `FileHandleReadWrap` (happens at least in Http2) 

For the first one I created https://github.com/nodejs/node/pull/27581 to avoid the reuse. Depending on the result of this one I can create one more to avoid the reuse of `FileHandleReadWrap`.

**Comment by mmarchini** [comment]
@Flarna is the FileHandleReadWrap still reused? If so I suggest creating a separate issue for it on nodejs/core and closing this one.

**Comment by Flarna** [comment]
Reuse of FileHandleReadWrap should be fixed by https://github.com/nodejs/node/pull/31972
